Add AsyncWriteExt::write_all_vectored utility

Thomasdezeeuw opened the pull request:
Since vectored I/O has been stabilised I was missing this. Two things to note;

* this takes `&mut [IoSlice]`, where the call in `io::Read` takes `&[IoSlice]`. This is needed to advance the buffer inside the `IoSlice`.
* this uses some unsafe code, I couldn't get around without it.

TODO:
- [x] Docs.

Nemo157 commented on futures-util/src/io/write_all.rs:
```diff
@@ -40,3 +41,239 @@ impl<W: AsyncWrite + ?Sized + Unpin> Future for WriteAll<'_, W> {
+                            this.bufs[0].as_ptr().add(bytes_written),
+                            this.bufs[0].len() - bytes_written,
+                        )
+                    };
```
This seems to indicate a missing `IoSlice` API; either something like `into_slice` so that you could get the slice with correct lifetime back and use `split_at`, or having `split_at` directly on `IoSlice`. Would be good to open an issue about adding something to move this unsafety to `std`.

Thomasdezeeuw replied:
Or better yet and `advance` method that would do this for use.

Thomasdezeeuw replied:
Created https://github.com/rust-lang/rust/issues/62726.

Thomasdezeeuw replied:
This is now available using the `io_slice_advance` Rust feature, tracking issue: https://github.com/rust-lang/rust/issues/62726.

Thomasdezeeuw commented:
@Nemo157 I've created {IoSlice, IoSliceMut}::advance which does the most heavy lifting of this change in https://github.com/rust-lang/rust/pull/62987. However it's an unstable feature, now that `Future` is also available on stable how are nightly features supposed to be handled?

Thomasdezeeuw commented:
@Nemo157 sorry about the delay. I've rebased this on master, updated the code to use the newly created `io_slice_advance` Rust feature and added documentation.

Thomasdezeeuw commented on futures-util/src/io/mod.rs:
```diff
@@ -500,6 +502,61 @@ pub trait AsyncWriteExt: AsyncWrite {
+    ///
+    /// [vectored writes]: std::io::Write::write_vectored
+    ///
+    /// # Notes
```
Please carefully review this section, I don't if it clearly states what I mean.

Nemo157 replied:
_If_ the future successfully completes, `bufs` is an empty slice, right?

On failure `bufs` may have been modified, but is not guaranteed to match with what has actually been written out (because I believe the underlying IO doesn't guarantee that it either writes some data or returns an error, it could have written some of the data _and then_ returned an error). That's pretty much equivalent to what `write_all` guarantees, if it returns an error some part of the data will have been written with no way to know how much.

I think this should try to avoid the "undefined" term since that has such strong memory-safety connotations, and the failure cases are relatively well-defined.

Thomasdezeeuw replied:
> _If_ the future successfully completes, `bufs` is an empty slice, right?

That is correct, but in the future we could optimise to just return ready if the entire (remaining) buffer is written, not modifying `bufs`.

> On failure `bufs` may have been modified, but is not guaranteed to match with what has actually been written out (because I believe the underlying IO doesn't guarantee that it either writes some data or returns an error, it could have written some of the data _and then_ returned an error). That's pretty much equivalent to what `write_all` guarantees, if it returns an error some part of the data will have been written with no way to know how much.

Do you want this to be reflected in the documentation somehow, or leave it as is?

> I think this should try to avoid the "undefined" term since that has such strong memory-safety connotations, and the failure cases are relatively well-defined.

Hence I added the it was not about memory safety, but I do understand the concern. Should I replace "undefined" with "unknown" (or something else) or you want to provide the guarantee that `bufs` will be empty after it successfully returns (also see my first remark in this comment)?

taiki-e commented on futures-util/src/io/write_all.rs:
```diff
@@ -33,10 +35,208 @@ impl<W: AsyncWrite + ?Sized + Unpin> Future for WriteAll<'_, W> {
+/// [`write_all_vectored`](super::AsyncWriteExt::write_all_vectored) method.
+#[derive(Debug)]
+#[must_use = "futures do nothing unless you `.await` or poll them"]
+#[cfg(feature = "write_all_vectored")]
```
Could you split these into a new file(`futures-util/src/io/write_all_vectored.rs`) instead of using cfg for each item?

Thomasdezeeuw replied:
Done.

Thomasdezeeuw commented:
Let me know if you want me to rebase this on the master branch.

Thomasdezeeuw commented:
@Nemo157 this pr has fallen out of date, but can I get an OK/not OK before spending more time on this?

Thomasdezeeuw commented:
@Nemo157 @taiki-e ping :)

Thomasdezeeuw commented:
@cramertj do you have time to review this?

cramertj commented:
Sure, sorry for the delay! Thanks for the ping.

cramertj commented on futures-util/Cargo.toml:
```diff
@@ -34,6 +34,7 @@ unstable = ["futures-core-preview/unstable"]
 cfg-target-has-atomic = ["futures-core-preview/cfg-target-has-atomic"]
 bilock = []
 read_initializer = ["io", "futures-io-preview/read_initializer", "futures-io-preview/unstable"]
+write_all_vectored = ["io"]
```
Can you add a corresponding feature to the top-level `futures` crate which turns this on, as well as a conditional re-export of the public types introduced?

cramertj commented on futures-util/src/io/mod.rs:
```diff
@@ -500,6 +505,61 @@ pub trait AsyncWriteExt: AsyncWrite {
+    ///
+    /// Different to `io::Write::write_vectored` this takes a *mutable*
+    /// reference to a slice of `IoSlice`s, not a non-mutable reference, because
+    /// we need to modify the slice to keep track of the bytes already written.
```
Who needs to? The future can store that state internally. Is this for the user to know? IMO we should try to match the API of `write_all` as closely as possible. Actually, for that matter, what we should be trying to match would be the yet-to-exist `std::io::Write::write_all_vectored`. If there isn't yet an issue for creating that API, would you mind creating one and linking it here? I'd really like to keep these extension traits from diverging from the API in the standard library, but I'm fine landing it here if it's already under a feature gate, so long as we make sure to tag that feature gate as `_unstable` / have it require an `unstable` feature flag in order to compile. Does that make sense?

Thomasdezeeuw replied:
The problem with writing all buffers is that is we need to deal with partially written `IoSlice`s. For which I wrote [`IoSlice::advance`](https://doc.rust-lang.org/std/io/struct.IoSlice.html#method.advance), which internally updates the slices, but needs mutable access. Basically `WriteAllVectored` needs ownership of `bufs`, to keep track of what bytes are already written. Also see the next section in the documentation.

In practice I don't think this will be a problem as `IoSlice` *slices*  are often created just before writing, e.g. see the example. But if you know of a better solution I'll love to hear it.

Thomasdezeeuw replied:
Created https://github.com/rust-lang/rust/issues/70436.

cramertj commented on futures-util/src/io/write_all_vectored.rs:
```diff
@@ -0,0 +1,200 @@
+use futures_core::future::Future;
```
Implementation and tests here look good to me, thanks!

cramertj commented on futures-util/src/io/mod.rs:
````diff
@@ -500,6 +505,61 @@ pub trait AsyncWriteExt: AsyncWrite {
+    /// # Examples
+    ///
+    /// ```
+    /// # futures::executor::block_on(async {
````
Thanks for including this example!

Thomasdezeeuw commented:
@cramertj I've rebased this on master and changed the docs to match `io::Write::write_all_vectored` in the std lib (as merged in #70612).

Thomasdezeeuw commented:
An unrelated tests failed `select_on_non_unpin_size` on the CI. @cramertj should I just increase the expected size in test? Or is this being tracked elsewhere?

Nemo157 commented:
Fixed in #2119, though that's still failing because of a nightly bug in the rustdoc check.

Thomasdezeeuw commented:
Now that #2119 is merged I've rebased this.

Thomasdezeeuw commented:
Friendly ping @cramertj, I don't want to have to rebase this again :)

cramertj commented:
Thanks for the ping, and for the changes! Sorry for the delay on my end.
